docs/architecture/adr-012-ibc-relayer-v2.md

```diff
@@ -0,0 +1,744 @@
+multiple specialized relayers that _prioritize_ differently on how packets
+should be relayed, such as based on incentivized packet fees or how long the
+packet has stayed idle.
+
+### Cosmos Relayer
+
+The all-in-one traits offered by the relayer framework serves as a starter
+pack for building custom relayers. To dogfood the all-in-one traits, we create
+the `ibc-relayer-cosmos` crate which offers the Cosmos-specific implementation
+of the relayer.
+
+Following the available all-in-one presets, the `relayer-cosmos` crate offers
+two variants of Cosmos chains, minimal and full, defined as follows:
+
+```rust
+pub struct MinCosmosChainContext {
+    pub handle: ChainHandle,
+    pub signer: Signer,
+    pub tx_config: TxConfig,
+    pub key_entry: KeyEntry,
+}
+
+pub struct FullCosmosChainContext {
+    pub handle: ChainHandle,
+    pub signer: Signer,
+    pub tx_config: TxConfig,
+    pub key_entry: KeyEntry,
+    pub batch_channel: CosmosBatchChannel,
+    pub telemetry: CosmosTelemetry,
+}
+```
+
+A visual inspection of the type definitions above shows that the full-featured
+Cosmos chain context contains two additional fields, `batch_channel` and
+`telemetry`. This is because the full preset makes use of the batched message
+sender and telemetry components, which requires the chain context to provide
+the batch channels and telemetry context. Hence, if a user want to use the
+full-featured Cosmos relayer, they would also have to instantiate and provide
+the additional parameters when constructing the chain context.
+
+The Cosmos chain context is implemented as an MVP for the relayer v1.5. As a
+result, it delegates most of the chain operations to the existing `ChainHandle`
+code in relayer v1. As we progress toward the v2 relayer, the Cosmos chain
+context is expected to eventually remove its dependency to `ChainHandle`, and
+directly holds the low-level fields such as `grpc_address`.
+
+To make use of the Cosmos chain context for relaying between Cosmos chains,
+we implement the one-for-all traits for our Cosmos chain contexts roughly as
+follows:
+
+```rust
+impl OfaBaseChain for MinCosmosChainContext
+{
+    type Preset = MinimalPreset;
+
+    type Error = CosmosError;
+
+    type Height = CosmosHeight;
+
+    type Timestamp = CosmosTimestamp;
+
+    type Message = CosmosMessage;
+
+    type Event = CosmosEvent;
+
+    type ChainStatus = CosmosChainStatus;
+
+    /* ... */
+
+    async fn query_chain_status(&self) -> Result<CosmosChainStatus, CosmosError> {
+        /* ... */
+    }
+
+    async fn send_messages(
+        &self,
+        messages: Vec<CosmosMessage>,
+    ) -> Result<Vec<Vec<CosmosEvent>>, CosmosError> {
+        /* ... */
+    }
+
+    /* ... */
+}
+```
+
+For demonstration purpose, the above code is slightly simplified from the actual
+Cosmos chain implementation. Readers are encouraged to refer to the
+`ibc-relayer-cosmos` itself to see the full implementation details.
```

Comment:
Let's link to the actual `impl OfaBaseChain for MinCosmosChainContext` code here.